test/runtests.jl

```diff
@@ -455,54 +473,547 @@ end
+
+    # TODO flow-sensitivity?
+    # ----------------------
+
+    let result = analyze_escapes((Any,Any)) do a, b
+            r = SafeRef{Any}(a)
+            r[] = b
+            return r[]
+        end
+        i = findfirst(isT(SafeRef{Any}), result.ir.stmts.type)::Int
+        r = findfirst(isreturn, result.ir.stmts.inst)::Int
+        @test_broken !has_return_escape(result.state.arguments[2], r) # a
+        @test has_return_escape(result.state.arguments[3], r) # b
+        @test is_sroa_eligible(result.state.ssavalues[i])
+    end
+    let result = analyze_escapes((Any,Any)) do a, b
+            r = SafeRef{Any}(:init)
+            r[] = a
+            r[] = b
+            return r[]
+        end
+        i = findfirst(isT(SafeRef{Any}), result.ir.stmts.type)::Int
+        r = findfirst(isreturn, result.ir.stmts.inst)::Int
+        @test_broken !has_return_escape(result.state.arguments[2], r) # a
+        @test has_return_escape(result.state.arguments[3], r) # b
+        @test is_sroa_eligible(result.state.ssavalues[i])
+    end
+    let result = analyze_escapes((Any,Any,Bool)) do a, b, cond
+            r = SafeRef{Any}(:init)
+            if cond
+                r[] = a
+                return r[]
+            else
+                r[] = b
+                return nothing
+            end
+        end
+        i = findfirst(isT(SafeRef{Any}), result.ir.stmts.type)::Int
+        r = findfirst(isreturn, result.ir.stmts.inst)::Int
+        @test has_return_escape(result.state.arguments[2], r) # a
+        @test_broken !has_return_escape(result.state.arguments[3], r) # b
+        @test is_sroa_eligible(result.state.ssavalues[i])
+    end
+end
+
+# demonstrate the power of our field / alias analysis with a realistic end to end example
+abstract type AbstractPoint{T} end
+mutable struct MPoint{T} <: AbstractPoint{T}
+    x::T
+    y::T
+end
+add(a::P, b::P) where P<:AbstractPoint = P(a.x + b.x, a.y + b.y)
+function compute(T, ax, ay, bx, by)
+    a = T(ax, ay)
+    b = T(bx, by)
+    for i in 0:(100000000-1)
+        a = add(add(a, b), b)
+    end
+    a.x, a.y
+end
+function compute(a, b)
+    for i in 0:(100000000-1)
+        a = add(add(a, b), b)
+    end
+    a.x, a.y
+end
+function compute!(a, b)
+    for i in 0:(100000000-1)
+        a′ = add(add(a, b), b)
+        a.x = a′.x
+        a.y = a′.y
+    end
+end
+let result = @analyze_escapes compute(MPoint, 1+.5im, 2+.5im, 2+.25im, 4+.75im)
+    for i in findall(isnew, result.ir.stmts.inst)
+        @test is_sroa_eligible(result.state.ssavalues[i])
+    end
+end
+let result = @analyze_escapes compute(MPoint(1+.5im, 2+.5im), MPoint(2+.25im, 4+.75im))
+    for i in findall(isnew, result.ir.stmts.inst)
+        @test is_sroa_eligible(result.state.ssavalues[i])
+    end
+end
+let result = @analyze_escapes compute!(MPoint(1+.5im, 2+.5im), MPoint(2+.25im, 4+.75im))
+    for i in findall(isnew, result.ir.stmts.inst)
+        @test is_sroa_eligible(result.state.ssavalues[i])
     end
 end
```

Comment:
So this PR will allow us to fully eliminate allocations involved within these examples (taken from [LuaJIT's allocation sinking](http://wiki.luajit.org/Allocation-Sinking-Optimization#implementation%5B)).